R/utils_render_rtf_redo.R

```diff
@@ -240,20 +240,181 @@ rtf_table <- function(rows) {
+    if (length(m)) {
+      as.numeric(m[[2]])
+    } else {
+      NA_real_
+    }
+  }, numeric(1))
+
+  units <- vapply(match, function(m) {
+    if (length(m)) {
+      m[[3]]
+    } else {
+      NA_character_
+    }
+  }, character(1))
+
+  # Check for negative values and stop of `allow_negative = FALSE`
+  non_na_vals <- vals[!is.na(vals)]
+  if (!allow_negative &&
+      length(non_na_vals) > 0 &&
+      any(!is.na(non_na_vals)) &&
+      is.numeric(non_na_vals) &&
+      any(non_na_vals < 0)) {
+
+    stop("Negative values supplied to widths cannot be used.",
+         call. = FALSE)
+  }
+
+  # Check for bad values and stop if necessary
+  bad_values <- nzchar(lengths_vec) & is.na(vals)
+  if (any(bad_values)) {
+
+    bad_values <- lengths_vec[nzchar(lengths_vec) & is.na(vals)]
+
+    stop(
+      "Some of the values supplied cannot be interpreted:\n",
+      "* Problem values are: ",
+      str_catalog(bad_values, surround = c("\"")), "\n",
+      "* Use either of: `px`, `pt`, `in`, `cm`, `mm`, or `tw` ",
+      "(e.g., \"12px\")",
+      call. = FALSE
+    )
+  }
+
+  dplyr::tibble(value = vals, unit = units)
+}
+
+abs_len_to_twips <- function(lengths_df) {
+
+  lengths_df %>%
+    dplyr::left_join(
+      tibble::enframe(twip_factors, name = "unit", value = "conv"),
+      by = c("unit" = "unit")
+    ) %>%
+    dplyr::mutate(
+      value = ifelse(!is.na(conv), value * conv, value),
+      unit = ifelse(!is.na(conv), "tw", unit)
+    ) %>%
+    dplyr::select(value, unit)
+}
+
+# The `col_width_resolver_rtf()` function returns a vector of
+# RTF column widths in units of twips (numeric)
+#
+# `table_width` and `col_widths` should be character vectors with
+# a numeric portion followed by the unit; the first should be a
+# length-1 vector and the latter should be a vector with as many
+# elements as there are visible columns in the gt table
+col_width_resolver_rtf <- function(table_width,
+                                   col_widths,
+                                   n_cols) {
+
+  rtf_page_width <- getOption("gt.rtf_page_width")
+
+  stopifnot(length(table_width) == 1)
+
+  if (table_width == "auto") {
+    table_width <- paste0(rtf_page_width, "tw")
+  }
+
+  if (is.null(col_widths)) {
+    col_widths <- rep_len("", n_cols)
+  }
+
+  table_width <- abs_len_to_twips(parse_length_str(table_width))
+
+  # For a table width set as a percentage, convert it to twips
+  # using the assumed width of a table set in a page
+  if (table_width$unit == "%") {
+
+    table_width$value <- (table_width$value / 100) * rtf_page_width
+    table_width$unit <- "tw"
+  }
+
+  # Ensure that the `table_width` unit is `"tw"`
+  stopifnot(isTRUE(table_width$unit == "tw"))
+
+  # If given a table width (i.e., not `"auto"`) without any specification
+  # of column widths, generate evenly spaced column widths based on the
+  # table width
+  if (is.null(col_widths)) {
+    col_widths <- paste0(rep(table_width$value / n_cols, n_cols), "tw")
+  }
```

Comment:
Remove this block--it's already handled above

Reply:
Now removed.